src/Controller/ControllerFactory.php

```diff
@@ -73,9 +96,23 @@ public function invoke($controller): ResponseInterface
         if ($result instanceof ResponseInterface) {
             return $result;
         }
-
         $action = $controller->getAction();
-        $args = array_values($controller->getRequest()->getParam('pass'));
+
+        $args = [];
+        $reflection = new ReflectionFunction($action);
+        $passed = array_values((array)$controller->getRequest()->getParam('pass'));
+        foreach ($reflection->getParameters() as $i => $parameter) {
+            if (isset($passed[$i])) {
```

Comment:
Checking for passed args first means the action methods would required declaring params for passed args before and additional params at the end. Wouldn't most people prefer the having params for passed args at the end instead? One reason being passed args could be optional. If for e.g. your method needs a service class it's unlikely to be optional.

Reply:
Yes that is a possibility too. I could make this argument handling code more robust and check if each argument exists in the container and if not, then draw from a passed argument until we run out of passed arguments.